File: omnibook-broker/src/main/java/com/sprint/omnibook/broker/api/exception/BrokerException.java
package com.sprint.omnibook.broker.api.exception;

import lombok.Getter;

import java.util.Map;

/**
 * Broker 모듈의 기본 예외.
 * 모든 도메인별 예외는 이 클래스를 상속받는다.
 */
@Getter
public class BrokerException extends RuntimeException {

    private final ErrorCode errorCode;
    private final Map<String, Object> details;

    public BrokerException(ErrorCode errorCode) {
        super(errorCode.getDefaultReason());
        this.errorCode = errorCode;
        this.details = null;
    }

    public BrokerException(ErrorCode errorCode, String reason) {
        super(reason);
        this.errorCode = errorCode;
        this.details = null;
    }

    public BrokerException(ErrorCode errorCode, Map<String, Object> details) {
        super(errorCode.getDefaultReason());
        this.errorCode = errorCode;
        this.details = details;
    }

    public BrokerException(ErrorCode errorCode, String reason, Map<String, Object> details) {
        super(reason);
        this.errorCode = errorCode;
        this.details = details;
    }

    public BrokerException(ErrorCode errorCode, Throwable cause) {
        super(errorCode.getDefaultReason(), cause);
        this.errorCode = errorCode;
        this.details = null;
    }

    public BrokerException(ErrorCode errorCode, String reason, Throwable cause) {
        super(reason, cause);
        this.errorCode = errorCode;
        this.details = null;
    }
}

File: omnibook-broker/src/main/kotlin/com/sprint/omnibook/broker/api/exception/BrokerException.kt
package com.sprint.omnibook.broker.api.exception

/**
 * Broker 모듈의 기본 예외.
 * 모든 도메인별 예외는 이 클래스를 상속받는다.
 */
open class BrokerException : RuntimeException {

    val errorCode: ErrorCode
    val details: Map<String, Any>?

    constructor(errorCode: ErrorCode) : super(errorCode.defaultReason) {
        this.errorCode = errorCode
        this.details = null
    }

    constructor(errorCode: ErrorCode, reason: String) : super(reason) {
        this.errorCode = errorCode
        this.details = null
    }

    constructor(errorCode: ErrorCode, details: Map<String, Any>) : super(errorCode.defaultReason) {
        this.errorCode = errorCode
        this.details = details
    }

    constructor(errorCode: ErrorCode, reason: String, details: Map<String, Any>) : super(reason) {
        this.errorCode = errorCode
        this.details = details
    }

    constructor(errorCode: ErrorCode, cause: Throwable) : super(errorCode.defaultReason, cause) {
        this.errorCode = errorCode
        this.details = null
    }

    constructor(errorCode: ErrorCode, reason: String, cause: Throwable) : super(reason, cause) {
        this.errorCode = errorCode
        this.details = null
    }
}
